File: solutions/java/0003.java
class Solution {
  public int lengthOfLongestSubstring(String s) {
    int ans = 0;
    Map<Character, Integer> lastSeen = new HashMap<>();

    for (int i = 0, j = 0; i < s.length(); ++i) {
      char c = s.charAt(i);
      if (lastSeen.containsKey(c))
        j = Math.max(j, lastSeen.get(c) + 1);
      ans = Math.max(ans, i - j + 1);
      lastSeen.put(c, i);
    }

    return ans;
  }
}

File: solutions/python3/0003.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans = 0
        lastSeen = {}

        j = 0
        for i, c in enumerate(s):
            if c in lastSeen:
                j = max(j, lastSeen[c] + 1)
            ans = max(ans, i - j + 1)
            lastSeen[c] = i

        return ans
